Removed dead AI code

reaperrr opened the pull request:
Edit: Changed PR to remove unused code instead.

reaperrr commented:
Note: It is hard to tell how much this effectively helps compared to bleed, so this will require some more extensive testing.

abcdefg30 commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -98,6 +98,12 @@ public class BuildingCategories
 		public readonly int CheckForNewBasesDelay = 1500;
 
+		[Desc("Delay (in ticks) before deciding which enemy to attack first.")]
+		public readonly int ChooseInitialEnemyDelay = 750;
```
Can you elaborate a bit why you added this?
I mean, couldn't we just use `UpdateAggroInterva` only instead?

reaperrr replied:
I just implemented the intitial delay first and added the interval later. You're right, it's better to keep just the interval.

reaperrr commented:
Updated.

reaperrr commented:
Updated, changed `*Interval` to `*Delay` to avoid an upgrade rule later. As @GraionDilach mentioned in #9810, we need to rename Interval to Delay for full consistency.

abcdefg30 commented on OpenRA.Mods.Common/AI/HackyAI.cs:
```diff
@@ -559,6 +565,36 @@ public void Tick(Actor self)
+				leastLikedEnemies.Random(Random) : liveEnemies.FirstOrDefault();
+
+			// Bump the aggro slightly to avoid changing our mind
+			if (leastLikedEnemies.Count() > 1)
```
What happens when `leastLikedEnemies` is `null`?

reaperrr replied:
I'm not sure whether there are any circumstances under which `leastLikedEnemies` can be `null` as long as there are still `liveEnemies`. I assumed there aren't any, but I'm not 100% sure how `MaxByOrDefault` works.

But yes, if it's possible for `leastLikedEnemies` to be null, that case needs to be covered, and if not, the `var enemy` can be simplified.

reaperrr commented:
Updated:
- now covers the case when `leastLikedEnemies` is `null` but `enemy` is not
- moved `var deadEnemies` behind the `liveEnemies.Any()` check so it's only performed if there are any live enemies remaining

RoosterDragon commented:
So a quick breeze over HackyAI shows that `Aggro` is only used within `ChooseEnemyTarget` and `ChooseEnemyTarget` is unused!

Everything else seems to find `FindClosestEnemy` to decide what to attack, which doesn't care about aggro.

Looks like some refactoring may be in order!

pchote commented:
Well that would explain why the AI suddenly turned dumb a couple of releases ago.

reaperrr commented:
So the AI sittling idle sometimes is due to the closest enemy being too far away...

RoosterDragon commented:
There is no upper limit to the search radius for finding an enemy most of the time, so at a glance I would expect the AI to attack things across the map if it had to.

This will probably need some deeper investigation to determine why the AI idles.

GraionDilach commented:
I've seen the AI going against worms in D2k immediately instead of attacking each other - during reproducing that activity-stack-overflow report - so I'd guess the AI ends up aggroing the wrong player. Couldn't check it yet how this PR helps on that though.

RoosterDragon commented:
First bad commit is e0b7242f1b0dac2c282aadbb2ecbca5fb6f3d1e1 (#2736, #2600), which would've affected release-20130514.

Given we've been without aggro for so long, it may be more prudent to remove the dead code and fix the existing code as needed.

reaperrr committed:
Remove dead code from HackyAI

Has been unused for three years...

reaperrr commented:
> So a quick breeze over HackyAI shows that Aggro is only used within ChooseEnemyTarget and ChooseEnemyTarget is unused!

It's used when an actor is damaged, Aggro against the attack's owner is buffed by the damage in that case (unless that's somehow not working either, but my impression is that it works, the AI usually responds to being attacked).

Anyway, AI is not exactly my strong point so I changed this PR to just remove the dead `ChooseEnemyTarget()` code for now.

RoosterDragon commented:
Sure, being attacked updates the aggro value. But nothing ever reads that value and uses it! So it has no effect in the end. You may as well get rid of it entirely.

(Also, you should remove the "fixes" in the OP lest merging this closes those now)

obrakmann commented:
:+1:

reaperrr committed:
Remove the rest of the unused AI Aggro left-overs

reaperrr commented:
Updated, removed the remaining unused aggro left-overs.

pchote commented:
It seems a shame to remove this instead of fixing it, but removing is better than leaving it broken so :+1:

pchote commented:
(no changelog because this has no player- or modder-visible changes)
